### PR TITLE
WIP: Lag monitor tracks empty teams

### PR DESCRIPTION
**Pull request by ArchShaman**
I think this might work. I'm not sure. In playerchanged() is where all the commshare stuff is now. It might be a simple solution? I'm not sure.

Looking for major feedback here. :)

### CODE REVIEW

File: LuaRules/Gadgets/game_lagmonitor.lua [hunk 1]
```diff
@@ -372,6 +380,22 @@ function gadget:GameFrame(n)
 	end	-- if
 end
 
+function gadget:PlayerChanged(id)
+	local _,_,spectator,newteamid,allyid,_ = Spring.GetPlayerInfo(id)
+	if originalteams[id] then
+		if newteamid ~= originalteams[id] and not spectator then -- commshare has gone and done its job. make note of new team
+			if #Spring.GetPlayerList(originalteams[id],true) == 0 and GG.Lagmonitor_activeTeams[allyid][originalteams[id]] then -- team is now empty :O tell OD immediately!
+				GG.Lagmonitor_activeTeams[allyid][originalteams[id]] = false
```
Comment (sprunk) [inline LuaRules/Gadgets/game_lagmonitor.lua]: no longer valid, see https://github.com/ZeroK-RTS/Zero-K/commit/5d8902d30c5e3a6f0189bdf779bc9b829abf647e

File: LuaRules/Gadgets/game_lagmonitor.lua [hunk 2]
```diff
@@ -372,6 +380,22 @@ function gadget:GameFrame(n)
...
+				GG.Lagmonitor_activeTeams[allyid].count = GG.Lagmonitor_activeTeams[allyid].count - 1
```
Comment (sprunk) [inline LuaRules/Gadgets/game_lagmonitor.lua]: Changing team does not change the count total of the allyteam

File: LuaRules/Gadgets/game_lagmonitor.lua [hunk 3]
```diff
@@ -372,6 +380,22 @@ function gadget:GameFrame(n)
 	end	-- if
 end
 
+function gadget:PlayerChanged(id)
+	local _,_,spectator,newteamid,allyid,_ = Spring.GetPlayerInfo(id)
+	if originalteams[id] then
+		if newteamid ~= originalteams[id] and not spectator then -- commshare has gone and done its job. make note of new team
+			if #Spring.GetPlayerList(originalteams[id],true) == 0 and GG.Lagmonitor_activeTeams[allyid][originalteams[id]] then -- team is now empty :O tell OD immediately!
```
Comment (sprunk) [inline LuaRules/Gadgets/game_lagmonitor.lua]: GetPlayerList disregards AFK status

File: LuaRules/Gadgets/game_lagmonitor.lua (hunk 2 above)
Comment (ArchShaman) [inline LuaRules/Gadgets/game_lagmonitor.lua]: Oh, so the count here isn't the team count but rather player count?

Comment (sprunk) [inline LuaRules/Gadgets/game_lagmonitor.lua]: It's a count of how many parts to split OD into.

Technically it's team count currently but the goal is for the shares to be based on players (so a team with N players gets N shares) so ultimately it should become player count.

Comment (ArchShaman) [inline LuaRules/Gadgets/game_lagmonitor.lua]: Thanks for the explanation. I've removed this in 89616e6 and simply got the teamsize for both teams to be the number of shares they get. Now the issue is the AFK players factoring in for both teams. 

Is it desirable to do teamsize - afkers or setting it to 0 if all players are afk?

File: LuaRules/Gadgets/game_lagmonitor.lua [hunk 4]
```diff
@@ -372,6 +380,18 @@ function gadget:GameFrame(n)
 	end	-- if
 end
 
+
+function gadget:PlayerChanged(id)
+	local _,_,spectator,newteamid,allyid,_ = Spring.GetPlayerInfo(id)
+	if originalteams[id] then
+		if newteamid ~= originalteams[id] and not spectator then -- commshare has gone and done its job. make note of new team sizes
+			GG.Lagmonitor_activeTeams[allyid][originalteams[id]] = #Spring.GetPlayerList(originalteams[id],true)
```
Comment (sprunk) [inline LuaRules/Gadgets/game_lagmonitor.lua]: Spring.GetPlayerList considers AFKers. So you'd need to iterate over the lagger list and make sure to exclude them. Note that the player being changed can be AFK as well (if a widget accepts the invite on his behalf).

Personally I'd do it someway like this:
```
function gadget:PlayerChanged(playerID)
    local _,_, spectator, newTeamID, allyID = Spring.GetPlayerInfo(playerID)
    local oldTeamID = originalteams[playerID]

    if laggers[playerID] -- no need to do anything now, will get added to the new team when coming back from afk
    or not oldTeamID -- can happen if a spec joins with /cheat
    or spectator
    then
        return
    end

    GG.Lagmonitor_activeTeams[allyID][oldTeamID] = GG.Lagmonitor_activeTeams[allyID][oldTeamID] - 1
    GG.Lagmonitor_activeTeams[allyID][newTeamID] = GG.Lagmonitor_activeTeams[allyID][newTeamID] + 1 
end
```

Note that regardless of method the afker handling function (in GameFrame) needs to be changed as well.

Comment (ArchShaman) [inline LuaRules/Gadgets/game_lagmonitor.lua]: I've started to rewrite the gameframe stuff. I do have a question:
Do you want teams to have share equal to their playercount as long as >=1 player is active or do you want afkers to not contribute their od shares?

Comment (sprunk) [inline LuaRules/Gadgets/game_lagmonitor.lua]: AFKers should not contribute.

Comment (ArchShaman) [inline LuaRules/Gadgets/game_lagmonitor.lua]: Okay.

**Comment by ArchShaman** [comment]
Closing -- updating repo for optimization PR.

**Pull request closed without merging**